### DISCUSSION TITLE
About provider sign with TLS

### DISCUSSION THREAD

**Discussion by dingyaoyang** [openssl/openssl#23186] [Q&A]
Through the openssl.org and discussion and YouTube videos, I implemented one provider which be used to sign and verify and include signature and keymgmt. And for easy debugging, I output information to the console at the end or beginning of each function.

Here's how I use:

 The first version:  Provider load into global
```
    OSSL_PROVIDER_add_builtin(NULL,"myprov",OSSL_provider_init);
    OSSL_PROVIDER_load(NULL,"myprov");
    EVP_set_default_properties(NULL,"?provider=default");

    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(NULL,"SM2","provider=custom");
    EVP_PKEY* pkey = NULL;
    OPENSSL_assert(EVP_PKEY_fromdata_init(pctx));
    OPENSSL_assert(EVP_PKEY_fromdata(pctx, &pkey, EVP_PKEY_KEYPAIR, NULL));

    if(SSL_CTX_use_PrivateKey(ssl_ctx,pkey)){
        ERR_reason_error_string(ERR_get_error());
        return -1;
    }
```
Server output info:
```
In provider/provider.c[43]:OSSL_provider_init_finish
prov request 10 operation id
In provider/keymgmt.c[55]:kmt_new_finish
In provider/keymgmt.c[79]:kmt_import_finish
We are the server on port: 10012

Client TCP connection accepted
prov request 1 operation id
prov request 12 operation id
prov request 10 operation id
prov request 21 operation id
prov request 10 operation id
prov request 21 operation id
prov request 10 operation id
prov request 21 operation id
prov request 10 operation id
prov request 21 operation id
prov request 11 operation id
prov request 4 operation id
prov request 3 operation id
prov request 10 operation id
prov request 21 operation id
prov request 2 operation id
Client TLCP connection accepted

Received: 123
Received: 1234
```
"prov request x operation id" comes from OSSL_FUNC_PROVIDER_QUERY_OPERATION output. It's clear that no signature is called

Then here's second version:  Provider load into a separate libctx
```
    OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();
    OSSL_PROVIDER_add_builtin(libctx,"myprov",OSSL_provider_init);
    OSSL_PROVIDER_load(libctx,"myprov");
    EVP_set_default_properties(NULL,"?provider=default");

    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(libctx,"SM2","provider=custom");
    EVP_PKEY* pkey = NULL;
    OPENSSL_assert(EVP_PKEY_fromdata_init(pctx));
    OPENSSL_assert(EVP_PKEY_fromdata(pctx, &pkey, EVP_PKEY_KEYPAIR, NULL));

    if(SSL_CTX_use_PrivateKey(ssl_ctx,pkey)){
        ERR_reason_error_string(ERR_get_error());
        return -1;
    }
```
Server output info:
```
In provider/provider.c[43]:OSSL_provider_init_finish
prov request 10 operation id
In provider/keymgmt.c[55]:kmt_new_finish
In provider/keymgmt.c[79]:kmt_import_finish
We are the server on port: 10012

Client TCP connection accepted
Client TLCP connection accepted

Received: hihihi
Received: ok
```
It seems that nothing happened.

Does anyone know what's going on? I'm a little dizzy.

**Comment by dingyaoyang** [comment]
I tested it and it didn't seem to have any problems：
```
    OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();
    OSSL_PROVIDER_add_builtin(libctx,"myprov",OSSL_provider_init);
    OSSL_PROVIDER_load(libctx,"myprov");
    EVP_set_default_properties(NULL,"?provider=default");

    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(libctx,"SM2","provider=custom");

    EVP_PKEY* pkey = NULL;
    OPENSSL_assert(EVP_PKEY_fromdata_init(pctx));
    OPENSSL_assert(EVP_PKEY_fromdata(pctx, &pkey, EVP_PKEY_KEYPAIR, NULL));

    unsigned char sig[128] = "12345";
    size_t siglen = 5;
    pctx = EVP_PKEY_CTX_new_from_pkey(NULL, pkey,"provider=custom");
    OPENSSL_assert(EVP_PKEY_sign_init(pctx));
    OPENSSL_assert(EVP_PKEY_sign(pctx,sig,&siglen,"3",1));
```

output：
```
In provider/provider.c[43]:OSSL_provider_init_finish
prov request 10 operation id
In provider/keymgmt.c[55]:kmt_new_finish
In provider/keymgmt.c[79]:kmt_import_finish
kmt_query_op_name query 12 operation_id
In provider/keymgmt.c[74]:kmt_query_op_name_finish
prov request 12 operation id
In provider/signature.c[42]:sm2_newctx_finish
In provider/signature.c[53]:sm2_sign_init_finish
In provider/signature.c[63]:sm2_sign_finish
We are the server on port: 10012
```

**Reply by dingyaoyang** [reply]
But why wouldn't the TLS connection be signed with a custom provider?

-------
The inspiration for the second version: 
https://github.com/openssl/openssl/discussions/22139#discussioncomment-7049308

**Reply by mattcaswell** [reply]
A TLS connection will negotiate the signature algorithms to use based on what signature algorithms both the client and the server recognise. You seem to have implemented a provider supplying SM2. But default OpenSSL does not support SM2 as a sigalg - your provider would need to add it. IANA maintains a registry of sigalgs here:

https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-signaturescheme

There is a sigalg registered for SM2 on the above website so your provider would have to declare support for it via a capability (and both client and server would need support for it to be used). See the following video for further details:

https://www.youtube.com/watch?v=BEZ2CKJH27E

It's unclear from your question and code whether you are implementing a client or a server. Since you are supplying a key, I assume you are implementing a server. Did you also attempt to supply a corresponding certificate? You need to supply both as well as taking the steps I outline above.

**Reply by dingyaoyang** [reply]
Thanks for your reply.

Yes, Openssl does not support SM2, so I use Tongsuo  (https://github.com/Tongsuo-Project/Tongsuo) to do that.

It seems that I should implement a GROUPs. Tomorrow I will try it.

**Reply by mattcaswell** [reply]
Ah ok. I'm not familiar with what changes that project has made. Although it does look like they've added the relevant group:

https://github.com/Tongsuo-Project/Tongsuo/blob/f1bde125eefa3b3ec196ba28e073812d41affe75/providers/common/capabilities.c#L204-L206

**Reply by dingyaoyang** [reply]
> It's unclear from your question and code whether you are implementing a client or a server. Since you are supplying a key, I assume you are implementing a server. Did you also attempt to supply a corresponding certificate? You need to supply both as well as taking the steps I outline above.

Yes, it's a server output. I need to implement mutual authentication. So, client and server load full certificate chain, sign cert and enc cert, sign key and enc key. The public key algorithm of the certificate is SM2.

> Although it does look like they've added the relevant group:
>
> https://github.com/Tongsuo-Project/Tongsuo/blob/f1bde125eefa3b3ec196ba28e073812d41affe75/providers/common/capabilities.c#L204-L206

That's great. I create an OSSL_FUNC_provider_get_capabilities_fn in my provider, but it wasn't called.

**Reply by mattcaswell** [reply]
> I create an OSSL_FUNC_provider_get_capabilities_fn in my provider, but it wasn't called.

Your provider has to be loaded into the libctx associated with the SSL_CTX for this to happen.

**Reply by dingyaoyang** [reply]
How do I do it? Via a pkey?

```
    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(libctx,"SM2","provider=custom");

    EVP_PKEY* pkey = EVP_PKEY_CTX_get0_pkey(pctx);
    OPENSSL_assert(EVP_PKEY_fromdata_init(pctx));
    OPENSSL_assert(EVP_PKEY_fromdata(pctx, &pkey, EVP_PKEY_KEYPAIR, NULL));

    if(!SSL_CTX_use_PrivateKey(ssl_ctx,pkey)){
        printf("Error: %s\n", ERR_reason_error_string(ERR_get_error()));
        return -1;
    }
```

But the private key and the certificate are to be matched. This Separate key can't to be load.

**Reply by mattcaswell** [reply]
What libctx are you using when you create the SSL_CTX? Do you create the SSL_CTX via SSL_CTX_new() or SSL_CTX_new_ex()? If the former then you are using the default libctx and you must load your provider into that libctx. If you are using SSL_CTX_new_ex() then you must load your provider into whatever lbctx you specified.

**Reply by dingyaoyang** [reply]
I see. I use SSL_CTX_new() and I should load provider into NULL. But nothing happened.

**Reply by mattcaswell** [reply]
By "nothing happened" you mean your "get_capabilities" function was not called? Did you load the provider into the default libctx (NULL) before creating the SSL_CTX? Has your "get_capabilities" function been added to the output OSSL_DISPATCH table from your provider init function?

**Reply by dingyaoyang** [reply]
Worked when I used it before creating the SSL_CTX.
